Documentation update (#254)
omnivector-solutions/license-manager: docs-source/src/architecture.rst

============================
The License Manager is based on a client/backend architecture. The backend consists of a RESTful API built with Python over a
PostgreSQL database. The client consists of a timed reconcile job that runs on the control node in the cluster and
a prolog integration to Slurm. The License Manager may serve multiple Slurm clusters and account for license consumption
originating outside these Slurm clusters.

Licenses Manager concept
------------------------
The current Slurm (23.02) licence accounting capabilities are limited to handle only consumption occurring within the cluster
using counters. Hence, license consumption originating in other clusters or workstations is not accounted for by this counter.

> The counter is setup defining the total number of licenses available and the number of consumed and free licenses are

Reviewer: ```suggestion
The counter is set up by defining the total number of licenses available, and the number of consumed and free licenses are
```
See: https://grammarist.com/spelling/set-up-vs-setup/#:~:text=Setup%20is%20one%20word%20when,set%20up%20the%20computer%E2%80%9D).

calculated based on current running jobs.

> When jobs are submitted to the cluster the number of licenses expected to be consumed is stated which allows for Slurm

Reviewer: ```suggestion
When jobs are submitted to the cluster, a parameter is passed to Slurm to set the number of licenses that are expected to be consumed. This allows for Slurm
```

> to only schedule jobs when licenses are available. When a job starts which requested licenses, the available licenses are

Reviewer: ```suggestion
to only schedule jobs when licenses are available. When a job that requested licenses starts, the number of available licenses are
```

> reduced. If not sufficient licenses are available, the job is not deemed eligible for submission by Slurm's scheduler and

Reviewer: ```suggestion
reduced. If sufficient licenses are not available, the job is not deemed eligible for submission by Slurm's scheduler and
```

will be held back in the queue.

> In a situation where license consumption may arise outside a single Slurm cluster, i.e. multiple clusters or consumed by
> other consumers, the internal counter of Slurm is not sufficient for efficient cluster and license utilization. The License

Reviewer: ```suggestion
In a situation where license consumption may arise outside a single Slurm cluster (i.e. multiple clusters share a license pool or licenses are used by
consumers other than Slurm), the internal counter of Slurm is not sufficient for efficient cluster and license utilization. The License
```

> Manager may then assist in keeping track of license consumption to continuously update the license availability to account
> for licenses consumed globally as well as licenses for jobs which have been started but not yet consumed the licenses from
> the license servers perspective.

Reviewer: ```suggestion
Manager provides improved license bookkeeping. It keeps track of license consumption to continuously update the license availability. To do this, License Manager accounts for global license consumption by interfacing directly with license servers. It also tracks licenses that have been reserved for jobs that have been started but not yet consumed the licenses from the license servers.
```

Author: This section could be further improved to get the nuances correct of the behavior. I will try to rework further.

Each job submitted to Slurm will trigger the ``PrologSlurmctld`` script that makes a request to the ``License Manager API``
to book the needed licenses prior to the allocation of the job. The booking ensures that the licenses are available for the job

> to consume at the license server by taking into consideration the licenses booked for other jobs and the total license usage

Reviewer: ```suggestion
to consume from the license server by taking into consideration the licenses booked for other jobs and the total license usage
```

> in the license server. It also prevents jobs from being eligible to start when the numbers of licenses free from the licenses
> servers perspective is higher or equal to what has been requested by the job.

Reviewer: ```suggestion
in the license server. It also prevents jobs from being eligible to start when the numbers of licenses that are reported as available directly from the licenses
servers is higher or equal to what has been requested by the job.
```

**********
A job can take some time to check out the license from the license server after it is submitted to Slurm.
Thus, each license has a ``grace time`` period that is used to indicate how long a booking will be retained.

> After the ``grace time`` expires, the booking is deleted. If the ``grace time`` expires the job is assumed to have failed

Reviewer 2: The grace time indicates how long the job takes to check out the license from the license server.

Once the grace time expires, the booking is deleted because the license is already accounted for in the license server (which implies that the Slurm counter was updated with the new value from the license server).

It doesn't mean that the job has failed if the grace time expires. In case of job failure, the epilog will delete the booking.

Author: In the previous versions of the License Manager ,in what is now the reconciliation, a feature matched license usage from the license server and with the lead host. Bookings were also removed as a part of the epilog script running after the job exited. This allowed for removal of the booking as soon as the licenses were checked out or the job terminated. If not licenses can be blocked by reservations no longer valid. 

Could you please clarify the current behavior so that it is reflected in the documentation.

Reviewer 2: The grace time represents the approximate period of time that a job takes to checkout a license from the license server. After this period has passed, it's assumed that the job checked out the license from the license server and the booking is no longer required.

In the reconciliation we clean up the bookings by the grace time:
- check all jobs that are running using squeue
- get the bookings for each running job
- check if the grace time for the license booked has expired
- if it has, the booking is deleted

This creates two opportunities for a booking to be deleted: when the job finishes or when the grace time is expired, whichever comes first.

> After the ``grace time`` expires, the booking is deleted. If the ``grace time`` expires the job is assumed to have failed
> and deemed not to consume the requested licenses, hence the booking is deleted to allow for other jobs start. Note that

Reviewer: ```suggestion
After the ``grace time`` expires, the booking is deleted. If the ``grace time`` expires the job is assumed to have failed
and the requested licenses are returned to the available license pool. In this case, the booking is deleted to allow for other jobs to start. Note that
```

Author: I think this section need some more work to point out the details of the behavior. I will try to rephrase it further.

the job may still consume compute resources until the run-time expires and will from the cluster counter perspective consume

> licenses.

Reviewer: I'm trying to think of a way to reword this, because it seems a bit unclear. I'm wondering if others have some thoughts here.

Author: I think this is an important aspect to cover and please propose how it can expressed better.
